File: 04_Asynchrony/exercise/word_count.js
import { readFile } from 'node:fs/promises';

async function processBook(urlString) {
    try {
        const response = await fetch(urlString);
        const text = await response.text();
        const words = text.toLowerCase().split(/\W+/);
        const counts = new Map();
        for (const word of words) {
            if (word) {
                counts.set(word, (counts.get(word) || 0) + 1);
            }
        }
        return counts;
    } catch (e) {
        console.error('Failed to process book:', urlString, e.message);
        return new Map();
    }
}

async function main() {
    const urls = (await readFile('books.csv', 'utf8')).trim().split('\n').filter(line => line.trim());

    const promises = urls.map(url => processBook(url));
    const bookCounts = await Promise.all(promises);

    const globalCounts = new Map();
    for (const counts of bookCounts) {
        for (const [word, count] of counts) {
            globalCounts.set(word, (globalCounts.get(word) || 0) + count);
        }
    }

    const sortedWords = Array.from(globalCounts.entries())
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10);

    console.log('\nTop 10 most frequent words:');
    for (let i = 0; i < sortedWords.length; i++) {
        const [word, count] = sortedWords[i];
        console.log(`${i + 1}. ${word}: ${count}`);
    }
}

await main();

File: 03_Tasks/exercise/WordCount.java
import java.net.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

public class WordCount {

    public static void main(String[] args) throws Exception {

        List<String> urls = Files.readAllLines(Path.of("books.csv"));

        ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor();

        // Process books in parallel
        List<Future<HashMap<String, Integer>>> futures = new ArrayList<>();
        for (String url : urls) {
            Future<HashMap<String, Integer>> future = executor.submit(() -> processBook(url));
            futures.add(future);
        }

        // Merge all hashmaps to get global word count
        HashMap<String, Integer> globalCounts = new HashMap<>();
        for (Future<HashMap<String, Integer>> future : futures) {
            HashMap<String, Integer> bookCounts = future.get();
            for (Map.Entry<String, Integer> entry : bookCounts.entrySet()) {
                globalCounts.merge(entry.getKey(), entry.getValue(), Integer::sum);
            }
        }

        // Output the 10 most frequent words
        List<Map.Entry<String, Integer>> sortedWords = globalCounts.entrySet().stream()
            .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
            .limit(10)
            .collect(Collectors.toList());

        System.out.println("\nTop 10 most frequent words:");
        for (int i = 0; i < sortedWords.size(); i++) {
            Map.Entry<String, Integer> entry = sortedWords.get(i);
            System.out.println((i + 1) + ". " + entry.getKey() + ": " + entry.getValue());
        }

        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.MINUTES);
    }

    static HashMap<String, Integer> processBook(String urlString) throws Exception {
        URL url = URI.create(urlString).toURL();
        String text = new String(url.openStream().readAllBytes());
        String[] words = text.toLowerCase().split("\\W+");
        HashMap<String, Integer> counts = new HashMap<>();
        for (String word : words) {
            if (!word.isEmpty()) {
                counts.merge(word, 1, Integer::sum);
            }
        }
        return counts;
    }
}
